### DISCUSSION TITLE
Deep `....`rest operator

### DISCUSSION THREAD

**Discussion by danielbayley** [DanielXMoore/Civet#1238] [Ideas]
How about something like this?

~~~ coffee
array := [1, [2, 3], 4]
console.log [....array] # [1, 2, 3, 4]

unpack := (....args) => console.log unpack
unpack(1, [2, 3], 4) # [1, 2, 3, 4]

a := a: { b: 2 }
b := a: { c: 3 }
c := a: 1
console.log { ....a, ....b, ...c } # { a: 1, b: 2, c: 3 }
~~~

Perhaps by compiling to something using `.flat()`…

Also thinking maybe `....` would flatten one level, while `.....` could be `Infinity`?

**Comment by edemaine** [comment]
I could see `....array` as shorthand for `...array.flat()` (and similar magic for function parameters). There are several examples of this in the Civet parser. For example:
```js
  TopLevelSingleLineStatements:first NestedTopLevelStatements*:rest ->
    return [
      ...first.map(s => ["", ...s]),
      ...rest.flat(),
    ]
```

Is this clearer with `....rest` instead? I'm not totally sure. It might be easily misread. `......rest` (like two `...`s) is maybe more intuitive to me, though we might be getting a little silly notation-wise. What do others think?

I don't think it makes sense to flatten other objects.

**Comment by bbrk24** [comment]
Four consecutive dots already has a meaning: https://civet.dev/playground?code=ZiAuLi4u
